src/fides/api/ctl/migrations/versions/985d3756f908_add_role_based_foundation.py

```diff
@@ -0,0 +1,50 @@
+"""add_role_based_foundation
+
+Revision ID: 985d3756f908
+Revises: d65bbc647083
+Create Date: 2023-02-17 21:04:20.113779
+
+"""
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = "985d3756f908"
+down_revision = "d65bbc647083"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column(
+        "client",
+        sa.Column("roles", sa.ARRAY(sa.String()), server_default="{}", nullable=True),
+    )
+    op.alter_column(
+        "client", "scopes", existing_type=postgresql.ARRAY(sa.VARCHAR()), nullable=True
+    )
+    op.add_column(
+        "fidesuserpermissions", sa.Column("roles", sa.ARRAY(sa.String()), nullable=True)
+    )
+    op.alter_column(
+        "fidesuserpermissions",
+        "scopes",
+        existing_type=postgresql.ARRAY(sa.VARCHAR()),
+        nullable=True,
+    )
```

Comment:
With all of these fields being nullable a user could have no permissions at all. Do we handle this somehow?

Reply:
Good callout. If a user has no permissions then they just can't do anything, which might make for a weird experience, but it doesn't break anything. I could add a code-side check when setting user permissions that they need either/or.

Reply:
I agree, other than a weird experience I don't see anything breaking with this. I'm open to whatever everyone thinks is best here. If we aren't worried about the chance someone has no roles I don't have an issue, it seems like a low possibility they wouldn't assign any roles.

Reply:
OK current state is that defaults are empty lists and we still set "privacy request viewer" permissions as a default at the code level.  After the roles UI is in, we can make the default a "role" viewer code side.

In the off-chance someone does have an empty list of scopes and roles, I'm throwing a 403 on login.

Reply:
Sorry I hadn't refreshed my page before I saw your comment Paul!